### PR TITLE
Split build_modules from build_web_compilers

### LINKED ISSUE #904: Allow configuring DDC module type
Looking at the source code I see that `build_web_compilers` has hardcoded `amd` module type when building for web.

Is there a way to override this default behavior?

I'm working on some packages which are designed for NodeJS and this requires `common` module type.

### PR DESCRIPTION
**Pull request by pulyaevskiy**
@jakemac53 @natebosch @matanlurey 

This is a quick draft on what it might look like to split modules and summaries in its own package as suggested.

Opening this to get general feedback on the approach. Only got existing tests passing in both build_modules and build_web_compilers, haven't tested on real projects yet.

Please let me know if you see any red flags.

Fixes #904.

### CODE REVIEW

File: build_modules/test/build_test.dart [hunk 1]
```diff
@@ -0,0 +1,65 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
```
Comment (jakemac53) [inline build_modules/test/build_test.dart]: I expected this to just be a move, we should be able to delete this equivalent test in build_web_compilers (and its dependencies on json_serializable, source_gen, and json_annotation)

**Comment by jakemac53** [comment]
Oh - one thing is we will want to create a `build.yaml` in the new package that contains the builders which have moved from `build_web_compilers`.

**Comment by jakemac53** [comment]
Also it looks like the e2e_test package needs to have its manual build script updated (see [here](https://travis-ci.org/dart-lang/build/jobs/336793030))

File: build_modules/README.md [hunk 2]
```diff
@@ -0,0 +1,101 @@
+# build_modules
```
[the remark this answers is not in the document]
Comment (pulyaevskiy) [reply build_modules/README.md]: Tried my best here. Please let me know if anything is still missing.

**Comment by pulyaevskiy** [comment]
@jakemac53 Thanks for the quick feedback!

All comments should be addressed now. I did update e2e_example. The tests are still failing badly on my Macbook but I see Travis is much happier now.

**Comment by jakemac53** [comment]
@pulyaevskiy which tests are failing on your macbook? We don't do mac tests right now on travis because there are a very limited number of mac workers available and it takes forever to schedule them.

File: build_modules/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (jakemac53) [reply build_modules/README.md]: This looks good, we don't need much I don't think as there won't be many direct consumers.

**Comment by jakemac53** [comment]
One last thing, we should update the `build_web_compilers` to the next breaking version and update the changelog.

**Comment by pulyaevskiy** [comment]
@jakemac53 here is the full log of running `pub run test -r expanded`
[e2e_example_pub_run_test_output.log](https://github.com/dart-lang/build/files/1697025/e2e_example_pub_run_test_output.log)


It seems like the issue might be somewhere in `expectTests*` matchers...

**Comment by jakemac53** [comment]
@pulyaevskiy ok, is it only the e2e_tests that are failing? That isn't super surprising, probably worth fixing though.

**Comment by pulyaevskiy** [comment]
@jakemac53 Tests in both `build_modules` and `build_web_compilers` are passing for me (one test skipped with `Skip: Should only be run during presubmit` message`).

**Comment by pulyaevskiy** [comment]
> That isn't super surprising, probably worth fixing though (but not in this pr unless you think it introduced the failures somehow)

Let me switch to master and run e2e tests there...

**Comment by pulyaevskiy** [comment]
On master: `+19 ~1 -8: Some tests failed.`, looks the same.

I will give it a try though, but agree it would be better to do in a separate PR.

**Comment by natebosch** [comment]
At https://github.com/dart-lang/build/commit/8e77ebca63ca5f8ff64e5832c2b73f36a9c66f54 the e2e_example tests are passing for me on mac.

Dart VM version: 2.0.0-dev.20.0 (Mon Jan 29 16:25:01 2018 +0100) on "macos_ia32"

**Comment by natebosch** [comment]
Could be a difference in our pub solve. Could you upload your `pubspec.lock`?

**Comment by pulyaevskiy** [comment]
Thanks @natebosch , I was on `dev.19.0` of the SDK. Upgraded to `dev.20.0`. Running e2e tests on master now and it just finished with all tests passing.

Will run on my branch now and let you know how it went.

**Comment by pulyaevskiy** [comment]
Yes, looks good with changes in this PR as well:

image

Anything else I've missed?
There is obviously a path dependency in `build_web_compilers` but it can only be updated after `build_modules` is published, so not sure what I can do there.

**Comment by pulyaevskiy** [comment]
I noticed there is no `analysis_options.yaml` in build_web_compilers (and build_modules as it's a copy). Should I add one to both with `strong-mode: true`?

**Comment by matanlurey** [comment]
@pulyaevskiy They use the root repository's:

https://github.com/dart-lang/build/blob/master/analysis_options.yaml

(Having one per-package is not necessary. We should probably remove duplicates)

**Comment by jakemac53** [comment]
> I noticed there is no analysis_options.yaml in build_web_compilers (and build_modules as it's a copy). Should I add one to both with strong-mode: true?

Ya the one that is in the package is actually just the default one that we use for summaries/ddc, its important that we don't use the actual root packages analysis_options.yaml because you might have stricter checks set up than the original package did, which results in compilation errors.

File: build_modules/pubspec.yaml [hunk 3]
```diff
@@ -0,0 +1,33 @@
+name: build_modules
+version: 0.0.1
```
Comment (natebosch) [inline build_modules/pubspec.yaml]: [nit] can we start this at `0.1.0`? We've started trying to shift over to that for the first release.

Comment (pulyaevskiy) [reply build_modules/pubspec.yaml]: Done!

**Comment by pulyaevskiy** [comment]
Looks like `pub get` timed out on Travis.

**Comment by jakemac53** [comment]
> Looks like pub get timed out on Travis.

ya I restarted the task, just a flake

**Comment by jakemac53** [comment]
whoops put my thank you message in the commit message on accident haha, but thanks @pulyaevskiy !

**Comment by pulyaevskiy** [comment]
:) Thank you guys for all the help and and quick responses!

**Pull request merged**